Fortify: Add map area markers for ACEX Fortify Objects

Seb105 opened the pull request:
**When merged this pull request will:**
- Add map markers to ACEX Fortify placed static objects
- Configurable via CBA: Always, Never, Only show markers for friendlies
- Removed with placed object deletion

![107410_20210420231932_1](https://user-images.githubusercontent.com/65898127/115471134-66652780-a22f-11eb-8a9d-e72d6cabb997.png)
![107410_20210420231937_1](https://user-images.githubusercontent.com/65898127/115471151-6bc27200-a22f-11eb-9d87-6536101da536.png)

Seb105 committed:
Add map markers for ACE Fortify Objects

Seb105 committed:
Fix side handling

Seb105 committed:
Fix out of date comments

jonpas commented on addons/fortify/XEH_postInit.sqf:
```diff
@@ -7,6 +7,7 @@ if (isServer) then {
         TRACE_3("objectPlaced",_unit,_side,_object);
         private _jipID = [QGVAR(addActionToObject), [_side, _object]] call CBA_fnc_globalEventJIP;
         [_jipID, _object] call CBA_fnc_removeGlobalEventJIP; // idealy this function should be called on the server
+        if (GVAR(markObjectsOnMap) isNotEqualTo 0 && {_object isKindOf "Static"}) then {[_object] call FUNC(createObjectMarker)};
```
```suggestion
        if (GVAR(markObjectsOnMap) isNotEqualTo 0 && {_object isKindOf "Static"}) then {
            [_object] call FUNC(createObjectMarker)
        };
```

jonpas commented on addons/fortify/XEH_postInit.sqf:
```diff
@@ -63,3 +64,18 @@ GVAR(objectRotationZ) = 0;
+        {!isNull player},
+        {
+            params ["_object"];
+            if (GVAR(markObjectsOnMap) isEqualTo 1 || {[_object getVariable QGVAR(objectSide), side group player] call BIS_fnc_sideIsEnemy}) then {
```
Might prefer [`side1 getFriend side2 < 0.6`](https://community.bistudio.com/wiki/getFriend), overhead of BIS functions is usually not great.

jonpas commented on addons/fortify/functions/fnc_createObjectMarker.sqf:
```diff
@@ -0,0 +1,48 @@
+
+/*
+ * Author: Seb
+ * Creates a map marker for a created static object but only for sides friendly to the creator side
```
```suggestion
 * Creates a map marker for a created static object but only for sides friendly to the creator side.
```

jonpas commented on addons/fortify/functions/fnc_createObjectMarker.sqf:
```diff
@@ -0,0 +1,48 @@
+ * Example:
+ * _object call acex_fortify_fnc_createObjectMarker
+ *
+ * Public: Yes
```
This function doesn't do enough input checking to be considered as public API in my opinion.

Seb105 replied:
Apologies, I thought that Public meant a Global result.

jonpas replied:
No worries, that's why there are reviews. :)

jonpas commented on addons/fortify/functions/fnc_createObjectMarker.sqf:
```diff
@@ -0,0 +1,48 @@
+private _markerNameStr = format ["acex_fortify_marker_%1", _object];
+
+// Create marker, set alpha using global event
+private _marker = createMarker [_markerNameStr, _object];
```
```suggestion
private _marker = createMarkerLocal [_markerNameStr, _object];
```
Same optimization as you did with the other commands.

jonpas commented on addons/fortify/initSettings.sqf:
```diff
@@ -9,3 +9,17 @@
         2
     ]
 ] call CBA_settings_fnc_init;
+[
```
```suggestion

[
```

jonpas commented on addons/fortify/initSettings.sqf:
```diff
@@ -9,3 +9,17 @@
+    true,
+    {},
+    true
+] call CBA_settings_fnc_init;
```
```suggestion
] call CBA_fnc_addSetting;
```

Seb105 replied:
Same file is also using CBA_settings_fnc_init, isn't mixing & matching worse?

jonpas replied:
Can also change the others, here or in separate PR. addSetting is public API, init is old.

jonpas commented on addons/fortify/stringtable.xml:
```diff
@@ -155,6 +155,21 @@
+            <English>Create map markers</English>
+        </Key>
+        <Key ID="STR_ACEX_Fortify_markObjectsOnMapDesc">
+            <English>Controls when budget update hints are shown</English>
```
```suggestion
            <English>Controls when budget update hints are shown.</English>
```

jonpas commented on addons/fortify/functions/fnc_createObjectMarker.sqf:
```diff
@@ -0,0 +1,48 @@
+private _direction = getDir _object;
+
+// Marker name unique to this object
+private _markerNameStr = format ["acex_fortify_marker_%1", _object];
```
```suggestion
private _markerNameStr = format [QGVAR(marker_%1), _object];
```

jonpas commented on addons/fortify/XEH_postInit.sqf:
```diff
@@ -63,3 +64,18 @@ GVAR(objectRotationZ) = 0;
+            };
+        }, 
+        _object
+    ] call CBA_fnc_waitUntilAndExecute;
```
There must be a better way to do this.

Seb105 replied:
Gotta wait somehow for the player object to exist for JIP side checking.

Seb105 committed:
Fix stringtable

Seb105 committed:
Update fnc_createObjectMarker.sqf

Seb105 committed:
Update XEH_postInit.sqf

Seb105 committed:
Update fnc_createObjectMarker.sqf

Seb105 committed:
Fix side handling.

Seb105 committed:
settings_fnc_init -> fnc_addsetting

Seb105 committed:
Actually fix side handling

mharis001 commented on addons/fortify/XEH_postInit.sqf:
```diff
@@ -63,3 +66,21 @@ GVAR(objectRotationZ) = 0;
+        {
+            params ["_object"];
+            private _objectSide = _object getVariable QGVAR(objectSide);
+            private _playerSide = side group player;
```
What happens if the player changes side mid-mission or remote controls a unit?

Seb105 replied:
TeamSwitch EH + CBA_fnc_addPlayerEventHandler covers player side change or RC, I can add that.

I do not believe there is a way to detect side relation changes (i.e setFriend) without a constant check.

mharis001 replied:
Which `CBA_fnc_addPlayerEventHandler`? I do not see one added by this PR. Also, there is a CBA player event handler for `"group"`.

Seb105 replied:
Typo, I meant to say "could cover". I will be implementing those 2 player EHs to recalculate the desired alphas when the current unit/group changes.

I've done some digging and I still don't see a way to handle side relation changes.

mharis001 commented on addons/fortify/stringtable.xml:
```diff
@@ -155,6 +155,21 @@
             <Russian>Показывать всегда</Russian>
             <Turkish>Her Zaman Göster</Turkish>
         </Key>
+        <Key ID="STR_ACEX_Fortify_markObjectsOnMap">
+            <English>Create map markers</English>
+        </Key>
+        <Key ID="STR_ACEX_Fortify_markObjectsOnMapDesc">
+            <English>Create map area markers akin to terrain buildings when an ACEX static object is placed.</English>
+        </Key>
+        <Key ID="STR_ACEX_Fortify_markObjectsOnMapNone">
+            <English>Never</English>
+        </Key>
+        <Key ID="STR_ACEX_Fortify_markObjectsOnMapFriendly">
+            <English>For friendlies only</English>
+        </Key>
+        <Key ID="STR_ACEX_Fortify_markObjectsOnMapEveryone">
+            <English>For everyone</English>
+        </Key>
```
Title case for display names and "an ACEX static object" -> "a fortify object" or "a fortification".

Seb105 committed:
Use setMarkerAlphaLocal

Seb105 committed:
Add player side handling

Seb105 committed:
Introduce wait for marker creation

Object position and rotation can be slightly off when on a server because the object is moved after creation, so wait a little.

Seb105 committed:
Update stringtable.xml

Seb105 committed:
Remove debug text
